### PR TITLE
add create/start/stop/restart/kill integration test

### PR DESCRIPTION
**Pull request by jimmyxian**
add create/start/stop/restart/kill integration test
Signed-off-by: Xian Chaobo <xianchaobo@huawei.com>

### CODE REVIEW

File: test/integration/api.bats [hunk 1]
```diff
@@ -48,3 +48,50 @@ function teardown() {
        [ "${#lines[@]}" -eq  2 ]
        [[ "${lines[1]}" == *"false"* ]]
 }
+
+
+@test "docker create/start/stop/restart/kill should return success" {
+	start_docker 1
```
Comment (vieux) [inline test/integration/api.bats]: can you use more than one daemon, like 3
so we are sure that the calls go to the right node

Comment (abronan) [inline test/integration/api.bats]: @vieux 2 might be enough then right?

Comment (vieux) [inline test/integration/api.bats]: yes but I prefer 3 :p

Comment (abronan) [inline test/integration/api.bats]: :ghost:

File: test/integration/api.bats [hunk 2]
```diff
@@ -48,3 +48,50 @@ function teardown() {
...
+	swarm_manage
+	
+	#create
+	run docker_swarm create --name test_container busybox sleep 1000
+	[ "$status" -eq 0 ]
+	run docker_swarm ps -l
+	[ "$status" -eq 0 ]
+	[ "${#lines[@]}" -eq 2 ]
+	[[ "${#lines[1]}" == *"test_container"* ]]
```
Comment (abronan) [inline test/integration/api.bats]: You need to remove the `#` in front of `lines[1]` (for all the following tests too)

**Comment by abronan** [comment]
@jimmyxian Test fails on `[[ "${#lines[1]}" == *"test_container"* ]]` because of the comment above.

File: test/integration/api.bats [hunk 3]
```diff
@@ -48,3 +48,50 @@ function teardown() {
...
        [[ "${lines[1]}" == *"false"* ]]
 }
+
+
+@test "docker create/start/stop/restart/kill should return success" {
+	start_docker 1
+	swarm_manage
+	
+	#create
+	run docker_swarm create --name test_container busybox sleep 1000
```
Comment (chanwit) [inline test/integration/api.bats]: This should be `run docker_swarm create --name test_container busybox sh` to make the container run forever.

**Comment by chanwit** [comment]
This PR relates to #492.

**Comment by aluzzardi** [comment]
Thanks @jimmyxian, much appreciated!

I've created a skeleton in #628 - basically created an integration test for every API call we are missing (the implementation right now is `skip`).

Could you please rebase this on top of #628 once we merge it?

**Comment by jimmyxian** [comment]
@aluzzardi OK, i will have done half of the apis integration test. If all test ok locally, i will subbmit it.

**Comment by chanwit** [comment]
:+1: 
@jimmyxian please submit it at small amounts. It's difficult to review all of them at a time :)

**Comment by jimmyxian** [comment]
@chanwit OK. :)

**Comment by aluzzardi** [comment]
Thanks @jimmyxian !

As @chanwit said, would it be possible to split it up into one test per "action" (create, start, ...)?

In `api.bats` I've created empty tests (they only do skip) for every single command the CLI supports.

Ideally, could you move each one of those tests into its own test function?

Agreed that for now it's not very useful, but each of those tests is going to grow: For `create`, we might want to test with different flags and see if they work. For `start`, `stop`, `restart`, we might want to do a `docker ps` and make sure the status got updated.

**Comment by jimmyxian** [comment]
@aluzzardi OK, Thks for your advice.I will follow this.

File: test/integration/api.bats [hunk 4]
```diff
@@ -38,9 +38,17 @@ function teardown() {
 	skip
 }
 
-# FIXME
 @test "docker create" {
-	skip
+	start_docker 3
+	swarm_manage
+
```
Comment (aluzzardi) [inline test/integration/api.bats]: Could you do a `ps -qa` and check that `lines -eq 0` before doing the create?

File: test/integration/api.bats [hunk 5]
```diff
@@ -91,9 +99,20 @@ function teardown() {
...
 
-# FIXME
 @test "docker kill" {
-	skip
+	start_docker 3
+	swarm_manage
+	# run 
+	run docker_swarm run -d --name test_container busybox sleep 1000
+	[ "$status" -eq 0 ]
+
```
Comment (aluzzardi) [inline test/integration/api.bats]: Could you do a `ps -l` here and check the container is running before the kill?

File: test/integration/api.bats [hunk 6]
```diff
@@ -207,9 +242,19 @@ function teardown() {
...
 
-# FIXME
 @test "docker start" {
-	skip
+	start_docker 3 
+	swarm_manage
+	# create
+	run docker_swarm create --name test_container busybox sleep 1000
+	[ "$status" -eq 0 ]
+
```
Comment (aluzzardi) [inline test/integration/api.bats]: Could you check that the container is stopped here?

File: test/integration/api.bats [hunk 7]
```diff
@@ -217,9 +262,19 @@ function teardown() {
...
 
-# FIXME
 @test "docker stop" {
-	skip
+	start_docker 3
+	swarm_manage
+	# run 
+	run docker_swarm run -d --name test_container busybox sleep 500
+	[ "$status" -eq 0 ]
+
```
Comment (aluzzardi) [inline test/integration/api.bats]: Check if container is running here

**Comment by aluzzardi** [comment]
Few comments but looks great!

**Comment by chanwit** [comment]
@jimmyxian This looks promising! But as @aluzzardi pointed out, every test should have its exit criteria  rather that only checking status codes.

For example, when calling `docker stop`, we need to ensure that the container is really stopped by querying the engine again.

**Comment by jimmyxian** [comment]
@aluzzardi  OK, thanks for your advice..I will modify it.

File: test/integration/api.bats (hunk 6 above)
Comment (jimmyxian) [inline test/integration/api.bats]: After creating the container, the status of the new container is null. I will check the lines of "docker ps -l" and check the container's name

**Comment by jimmyxian** [comment]
ping @aluzzardi Modified it

File: test/integration/api.bats [hunk 8]
```diff
@@ -91,9 +104,27 @@ function teardown() {
...
+	[[ "${lines[1]}" == *"test_container"* ]]
+	[[ "${lines[1]}" == *"Up"* ]]
+
+	# kill
+	run docker_swarm kill test_container
+	[ "$status" -eq 0 ]
+	# verify
+	run docker_swarm ps -l
+	[ "${#lines[@]}" -eq 2 ]
+        [[ "${lines[1]}" == *"test_container"* ]]
```
Comment (vieux) [inline test/integration/api.bats]: can you please fix the indent here ?

File: test/integration/api.bats [hunk 9]
```diff
@@ -177,9 +208,27 @@ function teardown() {
...
+	[[ "${lines[1]}" == *"test_container"* ]]
+	[[ "${lines[1]}" == *"Up"* ]]
+
+	# restart
+	run docker_swarm restart test_container
+	[ "$status" -eq 0 ]
+	# verify
+	run docker_swarm ps -l
+	[ "${#lines[@]}" -eq 2 ]
+        [[ "${lines[1]}" == *"test_container"* ]]
```
Comment (vieux) [inline test/integration/api.bats]: same here

File: test/integration/api.bats [hunk 10]
```diff
@@ -207,9 +256,26 @@ function teardown() {
...
+	run docker_swarm ps -l
+	[ "${#lines[@]}" -eq 2 ]
+	[[ "${lines[1]}" ==  *"test_container"* ]]
+
+	# start
+	run docker_swarm start test_container
+	[ "$status" -eq 0 ]
+	run docker_swarm ps -l
+	[ "${#lines[@]}" -eq 2 ]
+        [[ "${lines[1]}" == *"test_container"* ]]
```
Comment (vieux) [inline test/integration/api.bats]: and here

Comment (jimmyxian) [inline test/integration/api.bats]: @vieux ...ok

**Comment by jimmyxian** [comment]
ping @vieux  Done to fix the format

File: test/integration/api.bats (hunk 6 above)
Comment (aluzzardi) [inline test/integration/api.bats]: Can you verify if that's the case with docker as well?

You might have found another bug in swarm!

**Comment by aluzzardi** [comment]
One very last comment - otherwise LGTM

**Comment by aluzzardi** [comment]
Hey, for some reason my comment is now showing up here.

I added a last comment to `docker start`

File: test/integration/api.bats (hunk 6 above)
Comment (jimmyxian) [inline test/integration/api.bats]: OK, i will check it. :)

Comment (jimmyxian) [inline test/integration/api.bats]: @aluzzardi Yeah, I checked it. The docker is same as that.

**Comment by jimmyxian** [comment]
@aluzzardi Yeah, I checked it. After creating the container, the status of the new container is null.

**Comment by aluzzardi** [comment]
Awesome! Thanks!

LGTM

**Pull request merged**